File: client/src/lib/types/workflow.ts
export type Workflow = {
  id: string;
  name: string;
  data: WorkflowData;
};

export type WorkflowData = {
  nodes: Array<WorkflowNode>;
  edges: Array<WorkflowEdge>;
};

export type WorkflowEdge = {
  id: string;
  source: string;
  target: string;
};

export type WorkflowNode = {
  id: string;
  type: WorkflowNodeType;
  position: Position;
  data: Record<string, unknown>;
};

// These enum values (except Default) needs to be exact match to the values on server as they are mapped in "workflowMapper.ts"
export enum WorkflowNodeType {
  RequestNode = "RequestNode",
  HeadersNode = "HeadersNode",
}

export type RequestNode = {
  url: string;
  requestType: RequestType;
  requestBody?: string;
} & WorkflowNode;

export type HeadersNode = {
  headers: Array<HttpHeader>;
} & WorkflowNode;

export type HttpHeader = {
  name: string;
  value: string;
};

export type Position = {
  x: number;
  y: number;
};

export enum RequestType {
  Get = "Get",
  Post = "Post",
  Put = "Put",
  Patch = "Patch",
  Delete = "Delete",
}

export type WorkflowIdentity = {
  id: string;
  name: string;
};

/* Node data factories */
function createRequestNode(
  data?: Partial<Omit<RequestNode, "type">>,
): RequestNode {
  return {
    id: "",
    type: WorkflowNodeType.RequestNode,
    position: {x: 0, y: 0},
    url: "",
    requestType: RequestType.Get,
    ...data,
  };
}

function createHeadersNode(
  data?: Partial<Omit<HeadersNode, "type">>,
): HeadersNode {
  return {
    id: "",
    type: WorkflowNodeType.HeadersNode,
    position: {x: 0, y: 0},
    headers: [],
    ...data,
  };
}

export function createWorkflowNode(
  nodeType: WorkflowNodeType,
  position?: Position
): RequestNode | HeadersNode {
  switch (nodeType) {
    case WorkflowNodeType.RequestNode:
      return createRequestNode({position: position});
    case WorkflowNodeType.HeadersNode:
      return createHeadersNode({position: position});
    default:
      throw new Error(`Unsupported node type '${nodeType}'`);
  }
}

File: server/Workflows/Core/Models/Workflow.cs
using System.Text.Json.Serialization;

namespace WfAssist.Workflows.Core.Models;

internal sealed class Workflow
{
    public Guid Id { get; init; }
    public string Name { get; private set; }
    public WorkflowData Data { get; private set; }

    public Workflow(Guid id, string name, WorkflowData data)
    {
        Id = id;
        Name = name;
        Data = data;
    }

    public void ChangeName(string newName) => Name = newName;
    public void ChangeData(WorkflowData data) => Data = data;
}

internal sealed record WorkflowData
{
    public IEnumerable<Node> Nodes { get; init; } = [];
    public IEnumerable<Edge> Edges { get; init; } = [];
}

internal sealed record Edge
{
    public required string Id { get; init; }
    public required string Source { get; init; }
    public required string Target { get; init; }
}

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(RequestNode), nameof(RequestNode))]
[JsonDerivedType(typeof(HeadersNode), nameof(HeadersNode))]
internal abstract record Node
{
    public required string Id { get; init; }
    public required Size Size { get; init; }
    public required Position Position { get; init; }
}

internal sealed record RequestNode : Node
{
    public required RequestType RequestType { get; init; }
    public required string Url { get; init; }
    public string? RequestBody { get; init; }
}

internal sealed record HeadersNode : Node
{
    public List<HttpHeader> Headers { get; init; } = [];
}
